File: hometask_9/Source_9.1/stack.cpp
#include "stack.h"

const int error = -2147483646;

void createStack(Stack &stack) {
	Stack *newStack = new Stack;
	newStack->top = nullptr;

	stack = *newStack;
	delete newStack;
}

StackElement* createStackElement(int value) {
	StackElement *newStackElement = new StackElement;

	newStackElement->value = value;
	newStackElement->next = nullptr;

	return newStackElement;
}

void clear(Stack &stack) {
	while (!isEmpty(stack))
		pop(stack);
}

bool isEmpty(Stack &stack) {
	return stack.top == nullptr;
}

void push(Stack &stack, int value) {
	StackElement *newStackElement = createStackElement(value);

	newStackElement->next = stack.top;
	stack.top = newStackElement;
}

int pop(Stack &stack) {

	if (isEmpty(stack))
		return error;

	int result = stack.top->value;

	StackElement *tempElement = stack.top;
	stack.top = stack.top->next;
	delete tempElement;

	return result;
}

int top(Stack &stack) {
	if (isEmpty(stack))
		return error;

	return stack.top->value;
}

File: semester_2/hometask_1/Source_1.2/Stack.java
package com.group144.khodko;

public class Stack {

    private class StackElement {
        int value;
        StackElement next;

        StackElement(int value, StackElement next) {
            this.value = value;
            this.next = next;
        }
    }

    private StackElement head = null;

    Stack() { }

    Stack(int value) {
        head = new StackElement(value, null);
    }

    public void push(int value) {
        head = new StackElement(value, head);
    }

    public int pop() {
        if (head == null)
            return -404;
        int result = head.value;
        head = head.next;
        return result;
    }

    public int top() {
        if (head == null)
            return -404;
        return head.value;
    }
}
